proposals/low-level-struct-improvements.md

```diff
@@ -50,248 +45,226 @@ readonly ref struct Span<T>
 
-```cs
-class ExistingScenarios
+```c#
+Span<T> CompatExample(ref int p)
 {
-    Span<T> CreateSpan<T>(ref T p)
-    {
-        // The implementation of this method is irrelevant. From the point of
-        // the consumer the returned value is always safe to return.
-        ... 
-    }
-
-    Span<T> Examples<T>(ref T p, T[] array)
-    {
-        // Legal today
-        return CreateSpan(ref p); 
+    // Okay
+    int local = 42;
+    return CreateSpan<int>(ref local);
 
-        // Legal today, must remain legal
-        T local = default;
-        return CreateSpan(ref local);
+    // Okay
+    Span<T> span = stackalloc int[1];
+    return CreateSpan<int>(ref span[0]);
 
-        // Legal for any possible value that could be used as an argument
-        return CreateSpan(...);
-    }
+    // Okay
+    int[] array = new int[42];
+    return CreatSpan(ref array[0]);
 }
 ```
 
-The reason that all of the above samples are legal is because in the existing design there is no way for the return `Span<T>` to store a reference to the input state of the method call. This is because the span safety rules explicitly depend on `Span<T>` not having a constructor which takes a `ref` parameter and stores it as a field. 
+All of the above `return` statements are legal by our existing rules because the return of `CreateSpan` is always *safe-to-escape* to the *calling method*. The challenge in this proposal, which comes up many times, is they **must** remain legal in the version of the language which implements these features and / or when `Span<T>` moves to using `ref` fields. The above patterns can legally exist today and cannot become errors when moving to a new language version. 
 
-```cs
-class ExistingAssumptions
+Further the design must ensure that once `ref` fields exist that implementations of `CreateSpan<T>` cannot suddenly begin capture the input arguments by `ref`. For example: 
+
+```c#
+Span<T> CreateSpan<T>(ref T parameter)
 {
-    Span<T> CreateSpan<T>(ref T p)
-    {
-        // !!! Cannot happen today !!!
-        // The existing span safety rules specifically call out that this method
-        // cannot exist hence they can assume all returns from CreateSpan are
-        // safe to return.
-        return new Span<T>(ref p);
-    }
+    // Error: if this works it breaks the above compat requirements because the implementation
+    // is now capturing `ref` state.
+    return Span<T>(ref parameter);
```

Comment:
i'm not sure if this is a great example.  since the ref is captured, but only in somethign passed back to the caller, this is safe right (and i'm not sure how it breaks a compat requirement "if this works").

Reply:
Let me try expanding that sample out a bit. Essentially getting through the parts that will work (calling the ctor) and the parts that won't (returning it).